k2/python/k2/fsa.py

```diff
@@ -227,92 +251,156 @@ def __setattr__(self, name: str, value: Any) -> None:
+                       str(self.arcs)))
+        return properties
+
+    @properties.setter
+    def properties(self, value) -> None:
+        """ Only supports setting self.property to None; any other
+        value should be set internally by writing to self.__dict__ directly."""
+        if value == None:
+            self.__dict__['_properties'] = None
+        else:
+            raise RuntimeError("""Currently we don't allow the .properties of an Fsa
+            to be set this way, except to None""");
+
+    @property
+    def properties_str(self) -> str:
+        return _k2.fsa_properties_as_str(self._properties)
+
+    @property
+    def requires_grad(self) -> bool:
+        return self.scores.requires_grad
+
+    @property
+    def grad(self) -> bool:
+        return self.scores.grad
+
+
+
+    def __getattr__(self, name: str) -> Any:
+        """
+        Note: for attributes that exist as properties, e.g.
+        self.labels, self.properties, self.requires_grad, we won't
+        reach this code because Python checks the class dict before
+        calling getattr.  The same is true for instance attributes
+        such as self.{_tensor_attr,_non_tensor_attr,_cache,_properties}
+
+        The 'virtual' members of this class are those in self._tensor_attr
+        and self._non_tensor_attr.
+        """
+        if name in self._tensor_attr:
             return self._tensor_attr[name]
         elif name in self._non_tensor_attr:
             return self._non_tensor_attr[name]
-        elif name in self._grad_cache:
-            return self._grad_cache[name]
-        elif name == 'properties':
-            return self._properties
-        elif name == 'properties_str':
-            return _k2.fsa_properties_as_str(self._properties)
-        elif name == 'requires_grad':
-            return self.scores.requires_grad
-        elif name == 'grad':
-            return self.scores.grad
+        elif name in self._cache:
+            return self._cache[name]
+
 
         raise AttributeError(f'Unknown attribute {name}')
 
     def __delattr__(self, name: str) -> None:
-        assert name not in ('arcs', 'scores', 'labels', 'properties',
-                            '_properties', '_grad_cache')
+        # We won't allow deletion of class attributes such as @property
+        # getters
+        assert name not in Fsa.__dict__
+        # ... or instance attributes such as self._tensor_attr or
+        # self._properties
+        assert name not in self.__dict__
 
         if name in self._tensor_attr:
             del self._tensor_attr[name]
         elif name in self._non_tensor_attr:
             del self._non_tensor_attr[name]
-        elif name in self._grad_cache:
-            del self._grad_cache[name]
-        else:
-            super().__delattr__(name)
-
-    def _update_cache(self, name: str, value: Any) -> None:
-        self._grad_cache[name] = value
+        elif name in self._cache:
+            del self._cache[name]
```

Comment:
What if all the above tests fail?

Should we raise an exception for deleting a non-existent attribute?

Reply:
OK will raixe